### ISSUE TITLE
Kubeadm join fails with Invalid/expired token error in CoreOS

### ISSUE THREAD

**Issue by alexcpn** [kubernetes/kubeadm#1978]
## Versions

**kubeadm version** (use `kubeadm version`):

cluster1-2 bin # kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:17:50Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Kubernetes version** (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:20:10Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:12:17Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}

- **Cloud provider or hardware configuration**:

CoreOS VM on OpenShift

- **OS** (e.g. from /etc/os-release):
CoreOS
- **Kernel** (e.g. `uname -a`):

Linux cluster1-1.novalocal 4.19.86-coreos #1 SMP Mon Dec 2 20:13:38 -00 2019 x86_64 Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz GenuineIntel GNU/Linux

- **Others**:

## What happened?
I was setting up a new cluster with kubeadm . 

**In master**
```
1.cat << EOF > /etc/docker/daemon.json
{
  "exec-opts": ["native.cgroupdriver=systemd"]
}
EOF

2 .systemctl restart docker

3. docker info | grep -i cgroup
Cgroup Driver: systemd

4. kubeadm init --pod-network-cidr=10.244.0.0/16

You should now deploy a pod network to the cluster.
Run "kubectl apply -f [podnetwork].yaml" with one of the options listed at:
  https://kubernetes.io/docs/concepts/cluster-administration/addons/

Then you can join any number of worker nodes by running the following on each as root:
Got the token
kubeadm join 192.168.0.10:6443 --token vw6rtu.yyy\
    --discovery-token-ca-cert-hash sha256:xxx

5. Installed flannel 
```
**In woker node**
```
1.Updated hostname in /etc/hosts
2. Updated docker Cgroup driver
3. cluster1-2 bin # kubeadm join 192.168.0.10:6443 --token vw6rtu.08xn7hqhn2hadvek     --discovery-token-ca-cert-hash sha256:9acfeca4e2d0e420ddf068b7a29f1647ff296211cd204516c3aae755453404ad --v=2
W1218 11:11:43.253700    2155 join.go:346] [preflight] WARNING: JoinControlPane.controlPlane settings will be ignored when control-plane flag is not set.
I1218 11:11:43.253779    2155 join.go:371] [preflight] found NodeName empty; using OS hostname as NodeName
I1218 11:11:43.253852    2155 initconfiguration.go:103] detected and using CRI socket: /var/run/dockershim.sock
[preflight] Running pre-flight checks
I1218 11:11:43.253977    2155 preflight.go:90] [preflight] Running general checks
I1218 11:11:43.254110    2155 checks.go:249] validating the existence and emptiness of directory /etc/kubernetes/manifests
I1218 11:11:43.254129    2155 checks.go:286] validating the existence of file /etc/kubernetes/kubelet.conf
I1218 11:11:43.254139    2155 checks.go:286] validating the existence of file /etc/kubernetes/bootstrap-kubelet.conf
I1218 11:11:43.254148    2155 checks.go:102] validating the container runtime
I1218 11:11:43.336428    2155 checks.go:128] validating if the service is enabled and active
I1218 11:11:43.422723    2155 checks.go:335] validating the contents of file /proc/sys/net/bridge/bridge-nf-call-iptables
I1218 11:11:43.422921    2155 checks.go:335] validating the contents of file /proc/sys/net/ipv4/ip_forward
I1218 11:11:43.423063    2155 checks.go:649] validating whether swap is enabled or not
I1218 11:11:43.423188    2155 checks.go:376] validating the presence of executable ip
I1218 11:11:43.423292    2155 checks.go:376] validating the presence of executable iptables
I1218 11:11:43.423379    2155 checks.go:376] validating the presence of executable mount
I1218 11:11:43.423455    2155 checks.go:376] validating the presence of executable nsenter
I1218 11:11:43.423540    2155 checks.go:376] validating the presence of executable ebtables
I1218 11:11:43.423620    2155 checks.go:376] validating the presence of executable ethtool
I1218 11:11:43.423688    2155 checks.go:376] validating the presence of executable socat
I1218 11:11:43.423767    2155 checks.go:376] validating the presence of executable tc
I1218 11:11:43.423848    2155 checks.go:376] validating the presence of executable touch
I1218 11:11:43.423948    2155 checks.go:520] running all checks
I1218 11:11:43.518141    2155 checks.go:406] checking whether the given node name is reachable using net.LookupHost
I1218 11:11:43.518527    2155 checks.go:618] validating kubelet version
I1218 11:11:43.580704    2155 checks.go:128] validating if the service is enabled and active
I1218 11:11:43.593413    2155 checks.go:201] validating availability of port 10250
I1218 11:11:43.593746    2155 checks.go:286] validating the existence of file /etc/kubernetes/pki/ca.crt
I1218 11:11:43.593811    2155 checks.go:432] validating if the connectivity type is via proxy or direct
I1218 11:11:43.593909    2155 join.go:441] [preflight] Discovering cluster-info
I1218 11:11:43.594077    2155 token.go:188] [discovery] Trying to connect to API Server "192.168.0.10:6443"
I1218 11:11:43.594961    2155 token.go:73] [discovery] Created cluster-info discovery client, requesting info from "https://192.168.0.10:6443"
I1218 11:11:43.604812    2155 token.go:191] [discovery] Failed to connect to API Server "192.168.0.10:6443": token id "vw6rtu" is invalid for this cluster or it has expired. Use "kubeadm token create" on the control-plane node to create a new valid token
I1218 11:11:48.605151    2155 token.go:188] [discovery] Trying to connect to API Server "192.168.0.10:6443"
I1218 11:11:48.605631    2155 token.go:73] [discovery] Created cluster-info discovery client, requesting info from "https://192.168.0.10:6443"
I1218 11:11:48.607687    2155 token.go:191] [discovery] Failed to connect to API Server "192.168.0.10:6443": token id "vw6rtu" is invalid for this cluster or it has expired. Use "kubeadm token create" on the control-plane node to create a new valid token
I1218 11:11:53.607925    2155 token.go:188] [discovery] Trying to connect to API Server "192.168.0.10:6443"
I1218 11:11:53.608882    2155 token.go:73] [discovery] Created cluster-info discovery client, requesting info from "https://192.168.0.10:6443"
I1218 11:11:53.611365    2155 token.go:191] [discovery] Failed to connect to API Server "192.168.0.10:6443": token id "vw6rtu" is invalid for this cluster or it has expired. Use "kubeadm token create" on the control-plane node to create a new valid token
```
This is not a problem with token expirty; and firewalls are open for 6443

Still in master node again

```
$ kubeadm token create --ttl 0 --print-join-command
W1218 11:19:12.284747    1975 validation.go:28] Cannot validate kube-proxy config - no validator is available
W1218 11:19:12.284786    1975 validation.go:28] Cannot validate kubelet config - no validator is available
kubeadm join 192.168.0.10:6443 --token uwa2zs.g4zibdv54mbcjs32     --discovery-token-ca-cert-hash sha256:9acfeca4e2d0e420ddf068b7a29f1647ff296211cd204516c3aae755453404ad
```
In node I want to join

```
kubeadm join 192.168.0.10:6443 --token uwa2zs.g4zibdv54mbcjs32     --discovery-token-ca-cert-hash sha256:9acfeca4e2d0e420ddf068b7a29f1647ff296211cd204516c3aae755453404ad --v=4

Same error

2 token.go:191] [discovery] Failed to connect to API Server "192.168.0.10:6443": token id "uwa2zs" is invalid for this cluster or it has expired. Use "kubeadm token create" on the control-plane node to create a new valid token
```
## What you expected to happen?
kubeadm join should work

## How to reproduce it (as minimally and precisely as possible)?
Steps given

**Comment by neolit123** [comment]
this is very odd, but we have users using the latest kubeadm and also e2e test signal where this works, so it must be something on your end.

- can your try to delete the cluster completely and re-create the primary node using "kubeadm init" then join more nodes?
- if you can't do that above can you print the contents of the `cluster-info` config map in the kube-public namespace?

/triage support

**Comment by alexcpn** [comment]
Actually I deleted the all three nodes once and repeated steps before posting here. I will check again what exactly is wrong, however logs are not really helpful here

**Comment by alexcpn** [comment]
core@cluster1-1 / $ ./opt/bin/kubectl cluster-info
Kubernetes master is running at https://192.168.0.10:6443
KubeDNS is running at https://192.168.0.10:6443/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.

**Comment by alexcpn** [comment]
I deleted the nodes and tried with CentOS VM and it worked. There is another thing I changed; Earlier my OpenStack security group had a default security group also enabled. I removed the default this time, and just put my custom rules (enable all for test). I feel it could be that default rules ? . Also this time I changed the FloatingIPs to a new set. Not sure which of these changes caused it.  Will check and update

![image](https://user-images.githubusercontent.com/1157251/71145564-17c50900-2248-11ea-8a70-125b95e1159b.png)

**Comment by neolit123** [comment]
> kubectl cluster-info

i meant the output of this command (obtains a config map):
`kubectl get cm -n kube-public cluster-info -o yaml`

> I feel it could be that default rules ? .

not sure, if you share them in the #kubeadm or #kubernetes-users channels on slack, someone might be able to help. we don't have e2e signal for specific Linux distros.

**Comment by alexcpn** [comment]
(sorry I deleted the cluster; I will close this and check later)

**Comment by choryuidentify** [comment]
It's a filesystem problem. [check this issue.](https://github.com/kubernetes/kubernetes/issues/86961)

**Comment by sarthak8081** [comment]
W0212 20:35:27.269506   17312 join.go:346] [preflight] WARNING: JoinControlPane.controlPlane settings will be ignored when control-plane flag is not set.
[preflight] Running pre-flight checks
	[WARNING IsDockerSystemdCheck]: detected "cgroupfs" as the Docker cgroup driver. The recommended driver is "systemd". Please follow the guide at https://kubernetes.io/docs/setup/cri/
error execution phase preflight: couldn't validate the identity of the API Server: invalid public key hash, expected "format:value"
To see the stack trace of this error execute with --v=5 or higher 

@ alexcpn plase help me

**Comment by kymcaris** [comment]
> W0212 20:35:27.269506 17312 join.go:346] [preflight] WARNING: JoinControlPane.controlPlane settings will be ignored when control-plane flag is not set.
> [preflight] Running pre-flight checks
> [WARNING IsDockerSystemdCheck]: detected "cgroupfs" as the Docker cgroup driver. The recommended driver is "systemd". Please follow the guide at https://kubernetes.io/docs/setup/cri/
> error execution phase preflight: couldn't validate the identity of the API Server: invalid public key hash, expected "format:value"
> To see the stack trace of this error execute with --v=5 or higher
> 
> @ alexcpn plase help me

hello, i am also experiencing the same issue^

**Comment by KeyToon9** [comment]
> > W0212 20:35:27.269506 17312 join.go:346] [preflight] WARNING: JoinControlPane.controlPlane settings will be ignored when control-plane flag is not set.
> > [preflight] Running pre-flight checks
> > [WARNING IsDockerSystemdCheck]: detected "cgroupfs" as the Docker cgroup driver. The recommended driver is "systemd". Please follow the guide at https://kubernetes.io/docs/setup/cri/
> > error execution phase preflight: couldn't validate the identity of the API Server: invalid public key hash, expected "format:value"
> > To see the stack trace of this error execute with --v=5 or higher
> > @ alexcpn plase help me
> 
> hello, i am also experiencing the same issue^

same problem. Have you solved it?

**Comment by AntonioGonz** [comment]
Hello folks, 

I had the same issue after running the command: _kubeadm join --token **<token>** k8smaster:6443 --discovery-token-ca-cert-hash **<hash>**_

```
W0807 05:17:41.413517    2285 join.go:346] [preflight] WARNING: JoinControlPane.controlPlane settings will be ignored when control-plane flag is not set.
[preflight] Running pre-flight checks
	[WARNING IsDockerSystemdCheck]: detected "cgroupfs" as the Docker cgroup driver. The recommended driver is "systemd". Please follow the guide at https://kubernetes.io/docs/setup/cri/
error execution phase preflight: couldn't validate the identity of the API Server: invalid public key hash, expected "format:value"
To see the stack trace of this error execute with --v=5 or higher
```

Issue is comming because **<hash>** value is not in the correct format, as the issue indicates, expected format is as follows  "format:value", and based on documentation we have to specify hash value in the following format "<hash-type>:<hex-encoded-value>", hash-type should be **sha256**, so, command must be as follows for example:

`kubeadm join --token 0t7xcp.2e04gterfqg3rgtk k8smaster:6443 --discovery-token-ca-cert-hash sha256:2fc612865k9s82d45ff076dd426a93a3f6842036ee36e64cd1e0052a332ea855`

For more info just CTRL+F the flag **--discovery-token-ca-cert-hash** and you will find the correct format: https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join/

**Comment by As9530272755** [comment]
I also encountered this problem

**Comment by jayunit100** [comment]
why is this closed?

**Comment by jayunit100** [comment]
Fix would be - if they send in a bunk join token, just print that it should be correct ? agree this is tricky because unless you run w/ log level 5 you dont see it..

```

                                                                                                         /usr/local/go/src/runtime/proc.go:203                                                                                                                                         runtime.goexit                                                                                                                                                                                /usr/local/go/src/runtime/asm_amd64.s:1357                                                                                                                                    PS C:\Users\jay>                                                                                                                                                                      PS C:\Users\jay> kubeadm.exe join 10.0.0.42:6443 --token n657pt.edesxg1ka5yzj508     --discovery-token-ca-cert-hash sha256:226b96 --ignore-preflight-errors=SystemVerification --v=5  W1211 13:17:31.401885    4316 join.go:346] [preflight] WARNING: JoinControlPane.controlPlane settings will be ignored when control-plane flag is not set.                             I1211 13:17:31.402729    4316 join.go:371] [preflight] found NodeName empty; using OS hostname as NodeName                                                                            I1211 13:17:31.406621    4316 initconfiguration.go:103] detected and using CRI socket: npipe:////./pipe/docker_engine                                                                 [preflight] Running pre-flight checks                                                                                                                                                 I1211 13:17:31.407513    4316 preflight.go:90] [preflight] Running general checks                                                                                                     I1211 13:17:31.418833    4316 checks.go:249] validating the existence and emptiness of directory \etc\kubernetes\manifests                                                            I1211 13:17:31.419510    4316 checks.go:286] validating the existence of file \etc\kubernetes\kubelet.conf                                                                            I1211 13:17:31.419510    4316 checks.go:286] validating the existence of file \etc\kubernetes\bootstrap-kubelet.conf                                                                  I1211 13:17:31.420563    4316 checks.go:102] validating the container runtime                                                                                                         I1211 13:17:31.617459    4316 checks.go:128] validating if the service is enabled and active                                                                                          I1211 13:17:31.618309    4316 checks.go:520] running all checks                                                                                                                               [WARNING SystemVerification]: this Docker version is not on the list of validated versions: 19.03.13. Latest validated version: 19.03                                         I1211 13:17:32.039618    4316 checks.go:406] checking whether the given node name is reachable using net.LookupHost                                                                   I1211 13:17:32.047869    4316 checks.go:618] validating kubelet version                                                                                                               I1211 13:17:32.138044    4316 checks.go:128] validating if the service is enabled and active                                                                                          I1211 13:17:32.138044    4316 checks.go:201] validating availability of port 10250                                                                                                    I1211 13:17:32.142320    4316 checks.go:286] validating the existence of file C:/etc/kubernetes/pki/ca.crt                                                                            I1211 13:17:32.142320    4316 checks.go:432] validating if the connectivity type is via proxy or direct                                                                               I1211 13:17:32.142320    4316 join.go:441] [preflight] Discovering cluster-info                                                                                                       expected a 32 byte SHA-256 hash, found 3 bytes                                                                                                                                        k8s.io/kubernetes/cmd/kubeadm/app/util/pubkeypin.(*Set).allowSHA256                                                                                                                           /workspace/anago-v1.18.0-rc.1.21+8be33caaf953ac/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/util/pubkeypin/pubkeypin.go:96              k8s.io/kubernetes/cmd/kubeadm/app/util/pubkeypin.(*Set).Allow                                                                                                                                 /workspace/anago-v1.18.0-rc.1.21+8be33caaf953ac/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/util/pubkeypin/pubkeypin.go:56              k8s.io/kubernetes/cmd/kubeadm/app/discovery/token.retrieveValidatedConfigInfo                                                                                                                 /workspace/anago-v1.18.0-rc.1.21+8be33caaf953ac/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/discovery/token/token.go:64                 k8s.io/kubernetes/cmd/kubeadm/app/discovery/token.RetrieveValidatedConfigInfo                                                                                                                 /workspace/anago-v1.18.0-rc.1.21+8be33caaf953ac/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/discovery/token/token.go:51                 k8s.io/kubernetes/cmd/kubeadm/app/discovery.DiscoverValidatedKubeConfig                                                                                                                       /workspace/anago-v1.18.0-rc.1.21+8be33caaf953ac/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/ku

```****

**Comment by jayunit100** [comment]
in general, i reopened because i dont think we fixed this: the kubeadm join token being incorrect should be clearly explained at a high logging level .

**Comment by neolit123** [comment]
requiring a fix for this is valid. PR is here, but it will only be available in 1.21:
https://github.com/kubernetes/kubernetes/pull/97290

> to reproduce this error, just run kubeadm join with a bad token, and youll see that the message is tricky to uncover

the error messages for bad token are fine:
`Invalid value: "anng.2059h34kq4jt7vxr": the bootstrap token is invalid`

this is about the discovery token CA cert hash.

after the above PR, it should look like this:

```
couldn't validate the identity of the API Server: invalid discovery token CA certificate hash: ... <reason for error> (supported formats)
```
